python/cudf/cudf/core/dataframe.py

```diff
@@ -200,34 +200,57 @@ def __init__(self, data=None, index=None, columns=None, dtype=None):
         """
         super().__init__()
 
+        if isinstance(columns, (Series, cudf.Index)):
+            columns = columns.to_pandas()
+
         if isinstance(data, ColumnAccessor):
-            self._data = data
             if index is None:
-                index = as_index(range(self._data.nrows))
-            self.index = as_index(index)
-            return None
+                index = as_index(range(data.nrows))
+            else:
+                index = as_index(index)
+            self._index = index
 
-        if isinstance(data, DataFrame):
-            self._data = data._data
-            self._index = data._index
-            self.columns = data.columns
-            return
+            if columns is not None:
+                self._data = data.copy(deep=False)
+                df = self._reindex_frame(columns=columns, deep=True,)
+                self._data = ColumnAccessor(
+                    data=df._data,
+                    multiindex=data.multiindex,
+                    level_names=data.level_names,
+                )
```

Comment:
I think some of this can be made cleaner by supporting an `inplace` argument to `_reindex` (or just make that method operate in-place, while the public `reindex` is out-of-place).

```suggestion
                self._data = data.copy(deep=False)  # why do we need to copy?
                self._reindex(columns=columns, inplace=True)
```

Ditto below where this pattern is repeated

Reply:
I guess the copy is necessary if we're copying from another Frame's ColumnAccessor.

Reply:
Added an `inplace` parameter to `_reindex` but in this specific case we might want to preserve if it is a multiIndex or not for which we would have to pass-through the previous objects attributes.

Reply:
Shouldn't the `_reindex` method preserve the multiindex metadata of `self`? Maybe I'm misunderstanding.

Sorry to iterate so much on this point, but the patterns we establish here will likely be reused a lot across the entire codebase, so I think we need to get it right.

Reply:
> Shouldn't the `_reindex` method preserve the multiindex metadata of `self`? Maybe I'm misunderstanding.

While I was performing similar pattern there were some buffer size mismatch errors. I'll take another stab at it make it cleaner

Reply:
Investigated why the buffer size mismatch is happening, it is because when there is a mimic inplace operation happening for column we are setting the base_mask before setting the base_children incase of string column, which would lead to incorrect base_size calculation hence moved up setting base_children before we set base_mask.